### PR TITLE
use QGIS API instead of GDAL API to create heatmaps

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Rough attempt to port heatmap code to QGIS raster API. Based on @wonder-sk excellent work on #4022.

Heatmaps, created by new code are almost the same as original, very small difference caused by rounding and slightly different approach to define pixel size in QGIS API, probably this can be addressed somehow. As Processing test for Heatmap algorithm passes, difference should be a really small.

@nyalldawson, @wonder-sk I would be grateful for your review and opinions about this.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/analysis/raster/qgskde.cpp [hunk 1]
```diff
@@ -90,16 +83,35 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::prepare()
...
-    return FileCreationError;
-  mRasterBandH = GDALGetRasterBand( mDatasetH, 1 );
-  if ( !mRasterBandH )
+  if ( !mProvider->setNoDataValue( 1, NO_DATA ) )
+  {
     return FileCreationError;
+  }
+
+  int dataTypeSize = QgsRasterBlock::typeSize( Qgis::Float32 );
+  float *line = new float[ dataTypeSize * cols ];
```
Comment (wonder-sk) [inline src/analysis/raster/qgskde.cpp]: I think this should read ```new float[ cols ]```

File: src/analysis/raster/qgskde.cpp [hunk 2]
```diff
@@ -164,13 +176,14 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::addFeature( const
...
     // get the data
-    float *dataBuffer = ( float * ) CPLMalloc( sizeof( float ) * blockSize * blockSize );
-    if ( GDALRasterIO( mRasterBandH, GF_Read, xPosition, yPosition, blockSize, blockSize,
-                       dataBuffer, blockSize, blockSize, GDT_Float32, 0, 0 ) != CE_None )
-    {
-      result = RasterIoError;
-    }
+    QgsRasterBlock *block = mProvider->block( 1, extent, blockSize, blockSize );
+    QByteArray blockData = block->data();
+    blockData.detach();
```
Comment (wonder-sk) [inline src/analysis/raster/qgskde.cpp]: This detach is probably not necessary as the detach should happen on the line below in data() call if necessary

File: src/analysis/raster/qgskde.cpp [hunk 3]
```diff
@@ -196,22 +209,24 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::addFeature( const
         dataBuffer[ pos ] += pixelValue;
       }
     }
-    if ( GDALRasterIO( mRasterBandH, GF_Write, xPosition, yPosition, blockSize, blockSize,
-                       dataBuffer, blockSize, blockSize, GDT_Float32, 0, 0 ) != CE_None )
+
+    block->setData( QByteArray::fromRawData( ( char * )dataBuffer, sizeof( float ) * blockSize * blockSize ) );
```
Comment (wonder-sk) [inline src/analysis/raster/qgskde.cpp]: here I think ```block->setData( blockData );``` should be enough - and it would be easier to read...

**Review by wonder-sk — COMMENTED** [review]
Looks good to me - just minor comments. Thanks @alexbruy !

**Comment by nyalldawson** [comment]
I'm curious - what's driving this change? Is there a benefit to this approach over using gdal directly?

**Comment by alexbruy** [comment]
No benefits right now. But if one day QGIS will support other (editable) raster providers, using QGIS API will allow us to create heatmaps without or with very little additional code.

This is an experiment, I understand, that it does not bring any big improvement and I don't insist on merging this. I was curious how difficult will be to use QGIS API to create rasters from scratch, and here is the result.

**Comment by nyalldawson** [comment]
The future possibilities sound great :)

I'd have 2 concerns first though:
1. we'd need to make sure there's no speed regression by this flip (i.e. benchtest with a heatmap with many thousands/millions of points)
2. we'd need to make sure it handles (and doesn't crash!) with huge outputs the same/better then the current approach (i.e. a huge number of rows/cols in the output raster)

**Comment by alexbruy** [comment]
I will try to find suitable dataset with  thousands/millions of points and do some tests

**Comment by pcav** [comment]
Il 21/07/2017 07:14, Alexander Bruy ha scritto:
> I will try to find suitable dataset with thousands/millions of points
> and do some tests

why not using a randomly generated one?

-- 
Paolo Cavallini - www.faunalia.eu
QGIS & PostGIS courses: http://www.faunalia.eu/training.html
https://www.google.com/trends/explore?date=all&geo=IT&q=qgis,arcgis

**Comment by nyalldawson** [comment]
@alexbruy I think you'll need to pull in https://github.com/qgis/QGIS/commit/2f0c9d955b1ea497ea14df4e55bf3d640991302d too - my earlier fix flipped the heatmaps verticall!

**Comment by alexbruy** [comment]
I tested heatmap creation in QGIS release build using point layer with 500000 features (radius=0.00005, pixel_size=0.00001), full command
```
processing.run("qgis:heatmapkerneldensityestimation", {'INPUT':'/tmp/points500k.shp','RADIUS':0.00005,'PIXEL_SIZE':0.00001,'KERNEL':0,'DECAY':0,'OUTPUT_VALUE':0,'OUTPUT':QgsProcessingOutputLayerDefinition('/tmp/heatmap.tif')})
```
This generates 8.55 Gb raster file with dimensions 55466x41403. From my tests code with QGIS native API is about 6% slower then code using GDAL API.

Run  | GDAL API | QGIS API
------ | -------------- | -------------
1 | 2008019  | 2075438
2 | 1922636  | 2189702
3 | 2028768	 | 2075488
mean | 1986474  | 2113542

Probably this is because using QGIS API we need to pass extent of the area which then transformed into row/column representation, while in GDAL API we use row/column coordinates and don't need additional calculations.

File: src/analysis/raster/qgskde.cpp [hunk 4]
```diff
@@ -90,16 +83,35 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::prepare()
...
-  mRasterBandH = GDALGetRasterBand( mDatasetH, 1 );
-  if ( !mRasterBandH )
+  if ( !mProvider->setNoDataValue( 1, NO_DATA ) )
+  {
     return FileCreationError;
+  }
+
+  int dataTypeSize = QgsRasterBlock::typeSize( Qgis::Float32 );
+  float *line = new float[ cols ];
+  for ( int i = 0; i < cols; i++ )
```
Comment (m-kuhn) [inline src/analysis/raster/qgskde.cpp]: I guess `memset` eould do the same but more efficient?

File: src/analysis/raster/qgskde.cpp [hunk 5]
```diff
@@ -164,13 +176,13 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::addFeature( const
...
     // get the data
-    float *dataBuffer = ( float * ) CPLMalloc( sizeof( float ) * blockSize * blockSize );
-    if ( GDALRasterIO( mRasterBandH, GF_Read, xPosition, yPosition, blockSize, blockSize,
-                       dataBuffer, blockSize, blockSize, GDT_Float32, 0, 0 ) != CE_None )
-    {
-      result = RasterIoError;
-    }
+    QgsRasterBlock *block = mProvider->block( 1, extent, blockSize, blockSize );
+    QByteArray blockData = block->data();
+    float *dataBuffer = ( float * ) blockData.data();
```
Comment (m-kuhn) [inline src/analysis/raster/qgskde.cpp]: can this be changed to `static_cast` to be more cpp style?

File: src/analysis/raster/qgskde.cpp [hunk 6]
```diff
@@ -90,16 +83,35 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::prepare()
...
-    return FileCreationError;
-  mRasterBandH = GDALGetRasterBand( mDatasetH, 1 );
-  if ( !mRasterBandH )
+  if ( !mProvider->setNoDataValue( 1, NO_DATA ) )
+  {
     return FileCreationError;
+  }
+
+  int dataTypeSize = QgsRasterBlock::typeSize( Qgis::Float32 );
+  float *line = new float[ cols ];
```
Comment (m-kuhn) [inline src/analysis/raster/qgskde.cpp]: Can this use `std::vector` or `std::unique_ptr` to be less prone to leak memory?

Comment (alexbruy) [reply src/analysis/raster/qgskde.cpp]: Should be fixed now

File: src/analysis/raster/qgskde.cpp (hunk 4 above)
Comment (alexbruy) [reply src/analysis/raster/qgskde.cpp]: Using std::fill as we already use std::vector. Or better to use `memset` anyway?

File: src/analysis/raster/qgskde.cpp (hunk 5 above)
Comment (alexbruy) [reply src/analysis/raster/qgskde.cpp]: Hmm, I'm not a guru, but it looks like we can not use `static_cast` here without implementing custom conversion

Comment (m-kuhn) [reply src/analysis/raster/qgskde.cpp]: no worries, it was just a random thought. In general the new-style is more explicit about what's happening in the background.

File: src/analysis/raster/qgskde.cpp (hunk 4 above)
Comment (m-kuhn) [reply src/analysis/raster/qgskde.cpp]: Ah, right, sounds `std::fill` is the way to go (https://stackoverflow.com/questions/1373369/which-is-faster-preferred-memset-or-for-loop-to-zero-out-an-array-of-doubles#comment1217663_1373422).

**Comment by m-kuhn** [comment]
@alexbruy what is the status on this? do you think it's worth to resolve conflicts and merge?

**Comment by alexbruy** [comment]
@m-kuhn I will be happy to update this PR. But it does not brings any improvements right now, when we have only signle raster provider with editing capabilities (GDAL). Also when using QGIS API we have small slowdown caused by the additional calculations (row/col from extent).

**Comment by m-kuhn** [comment]
Should it rather be closed then?

**Comment by alexbruy** [comment]
Ok, let's close it. Probably one day we can return to this discussion.

**Comment by m-kuhn** [comment]
Congrats for bringing the pull request queue below 100 :medal_sports:
